packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2679,44 +2678,73 @@ export class ContainerRuntime
 				inboundBatch,
 				local,
 			);
+
 			if (messagesWithPendingState.length > 0) {
-				messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
-					const msg: MessageWithContext = {
-						message,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-						localOpMetadata,
-					};
-					this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
-				});
+				this.processBatch(
+					messagesWithPendingState,
+					local,
+					savedOp,
+					true /* modernRuntimeMessage */,
+				);
 			} else {
 				this.ensureNoDataModelChanges(() => this.processEmptyBatch(inboundBatch, local));
 			}
 		} else {
-			// Check if message.type is one of values in ContainerMessageType
-			// eslint-disable-next-line import/no-deprecated
-			if (isRuntimeMessage(messageCopy)) {
-				// Legacy op received
-				this.ensureNoDataModelChanges(() =>
-					this.processRuntimeMessage({
-						message: messageCopy as InboundSequencedContainerRuntimeMessage,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-					}),
-				);
-			} else {
-				// A non container runtime message (like other system ops - join, ack, leave, nack etc.)
-				this.ensureNoDataModelChanges(() =>
-					this.observeNonRuntimeMessage({
-						message: messageCopy as InboundSequencedNonContainerRuntimeMessage,
-						local,
-						isRuntimeMessage: false,
-						savedOp,
-					}),
-				);
-			}
+			const batch = [{ message: messageCopy, localOpMetadata: undefined }];
+			this.processBatch(batch, local, savedOp, undefined /* modernRuntimeMessage */);
+		}
+	}
+
+	/**
+	 * Processes a batch of messages. It emits "batchBegin" and "batchEnd" events and calls deltaScheduler.
+	 */
+	private processBatch(
+		batch: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata?: unknown;
+		}[],
+		local: boolean,
+		savedOp: boolean | undefined,
+		modernRuntimeMessage: true | undefined,
+	) {
+		const firstMessage = batch[0]?.message;
+		assert(firstMessage !== undefined, "Batch must have at least one message");
+		this.deltaScheduler.batchBegin(firstMessage);
+		this.emit("batchBegin", firstMessage);
+
+		let error: unknown;
+		try {
+			batch.forEach(({ message, localOpMetadata }) => {
+				const runtimeMessage = modernRuntimeMessage ?? isRuntimeMessage(message);
```

Comment:
nit: make `modernRuntimeMessage` a boolean and then 
```suggestion
				const runtimeMessage = modernRuntimeMessage || isRuntimeMessage(message);
```

Reply:
(Optional) Also, a thought - this code makes it seem like each message in the batch needs to check if it's a runtime message, as if you could have a mixed batch.  I understand the shared code and how some callers will pass a singleton array.  But someone else wouldn't.

Maybe move the runtimeMessage calculation outside the loop, something like this:

```typescript
const runtimeMessage = modernRuntimeMessage || batch.every(({ message }) => isRuntimeMessage(message));
assert(!runtimeMessage || batch.every(({ message }) => !isRuntimeMessage(message)), "Batches should be homogenous regarding runtime message or not");
```